src/wovnio/wovnphp/Url.php

```diff
@@ -49,89 +45,92 @@ public static function addLangCode($uri, $store, $lang, $headers)
+            if (!self::uriFromSameHost($no_lang_uri, $no_lang_host)) {
+                return $new_uri;
+            }
+            switch ($pattern) {
+                case 'subdomain':
+                    $new_uri = self::addSubdomainLangCode($parsed_url, $lang_code, $no_lang_uri);
+                    break;
+                case 'query':
+                    $new_uri = self::addQueryLangCode($no_lang_uri, $lang_code, $lang_param_name);
+                    break;
+                case 'path':
+                    $new_uri = self::addPathLangCode($no_lang_uri, $lang_code, $site_prefix_path);
+                    break;
+                default:
+                    $new_uri = $uri;
             }
+        } else {
+            // relative urls
 
-            $parsed_host = array_key_exists('host', $parsed_url) ? $parsed_url['host'] : null;
-            if ($parsed_host !== null && array_key_exists('port', $parsed_url)) {
-                $parsed_host = $parsed_host . ':' . $parsed_url['port'];
+            if (preg_match('/:/', $no_lang_uri)) {
+                // do nothing for protocols other than http and https (e.g. tel)
+                return $new_uri;
             }
 
-            // only continue if the host of the url is the same as the headers host
-            if ($parsed_host !== null && strtolower($parsed_host) === strtolower($no_lang_host)) {
-                switch ($pattern) {
-                    case 'subdomain':
-                        // check if subdomain already exists
-                        if (preg_match('/\./', $parsed_url['host'])) {
-                            $sub_do_a = explode('.', $parsed_url['host']);
-                            $sub_do = $sub_do_a[0];
-                            $sub_do_lang = Lang::getCode($sub_do);
-                            if ($sub_do_lang && strtolower($sub_do_lang) === strtolower($lang_code)) {
-                                $new_uri = preg_replace('/' . $lang_code . '/i', strtolower($lang_code), $no_lang_uri, 1);
-                            } else {
-                                $new_uri = preg_replace('/(\/\/)([^\.]*)/', '${1}' . self::formatForRegExp(strtolower($lang_code)) . '.' . '${2}', $no_lang_uri, 1);
-                            }
-                        } else {
-                            $new_uri = preg_replace('/(\/\/)([^\.]*)/', '${1}' . self::formatForRegExp(strtolower($lang_code)) . '.' . '${2}', $no_lang_uri, 1);
-                        }
-                        break;
-                    case 'query':
-                        $new_uri = self::addQueryLangCode($no_lang_uri, $lang_code, $lang_param_name);
-                        break;
-                    default:
-                        //path
+            switch ($pattern) {
+                case 'subdomain':
+                    $lang_url = $headers->protocol . '://' . strtolower($lang_code) . '.' . $headers->host;
```

Comment:
What is this code trying to do? I have no idea why subdomain pattern is dealing with paths at all.

Reply:
Didn't write this block of code, I only refactored them so they are less spaghetti-y. I think this block is trying to reconstruct a relative path back to a fully qualified path (protocol+host+path), then do the subdomain pattern thing on top.

Reply:
If we can prove this code doesn't actually do anything useful, it would be great to remove it. There really should not be much involved with add_lang_code for subdomain - add the lang code immediately after the protocol or before the first domain segment.

Reply:
My understanding is that `addLangCode` is trying to also account for cases when the input URI doesn't have its lang code removed. Maybe this is a behavior from older version where we don't explicitly "clean" the URIs. I will investigate this and if this behavior is no longer required, I'll remove the redundant code.

Reply:
For a future refactor though I agree. Just shocked at seeing some of the weird stuff the code is doing 😒

Reply:
Yep, we definitely can straighten things up more. Will resolve this convo for now.